File: backend/app/services/auth_service.py
from datetime import datetime, timedelta, timezone

from jose import JWTError, jwt
from app.config import get_settings
from app.schemas.auth import PublicUser
from app.schemas.health import HealthProfile
from app.security.passwords import hash_password, verify_password
from app.storage import users_store
from app.storage.errors import StorageConflict

DEFAULT_HEALTH_PROFILE = HealthProfile(
    ageRange="25-34",
    conditions=[],
    allergies=[],
    medications="",
)


class AuthError(Exception):
    def __init__(self, message: str, status_code: int = 400):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


def _to_public(record: dict) -> PublicUser:
    return PublicUser(
        id=record["id"],
        email=record["email"],
        name=record["name"],
        healthProfile=record["health_profile"],
        createdAt=record["created_at"],
    )


def _sign_token(user_id: str) -> str:
    settings = get_settings()
    expire = datetime.now(timezone.utc) + timedelta(days=7)
    payload = {"sub": user_id, "exp": expire}
    return jwt.encode(payload, settings.AUTH_SECRET, algorithm="HS256")


def verify_token(token: str) -> str:
    settings = get_settings()
    try:
        payload = jwt.decode(token, settings.AUTH_SECRET, algorithms=["HS256"])
        sub = payload.get("sub")
        if not sub or not isinstance(sub, str):
            raise AuthError("Invalid or expired session", 401)
        return sub
    except JWTError:
        raise AuthError("Invalid or expired session", 401)


def signup(email: str, password: str, name: str) -> tuple[PublicUser, str]:
    normalized = email.strip().lower()
    if users_store.find_by_email(normalized):
        raise AuthError("An account with this email already exists", 409)

    try:
        record = users_store.insert_user(
            email=normalized,
            password_hash=hash_password(password),
            name=name.strip(),
            health_profile=DEFAULT_HEALTH_PROFILE.model_copy(),
        )
    except StorageConflict as exc:
        raise AuthError("An account with this email already exists", 409) from exc
    return _to_public(record), _sign_token(record["id"])


def login(email: str, password: str) -> tuple[PublicUser, str]:
    normalized = email.strip().lower()
    record = users_store.find_by_email(normalized)
    if not record or not verify_password(password, record["password_hash"]):
        raise AuthError("Invalid email or password", 401)
    return _to_public(record), _sign_token(record["id"])


def get_user_by_id(user_id: str) -> PublicUser | None:
    record = users_store.find_by_id(user_id)
    return _to_public(record) if record else None


def update_user_profile(
    user_id: str,
    *,
    name: str | None = None,
    health_profile: HealthProfile | None = None,
) -> PublicUser:
    try:
        record = users_store.update_user(
            user_id,
            name=name.strip() if name is not None else None,
            health_profile=health_profile,
        )
    except StorageConflict as exc:
        raise AuthError(
            "Profile was modified elsewhere — please refresh and try again.",
            409,
        ) from exc
    if not record:
        raise AuthError("User not found", 404)
    return _to_public(record)


def clear_users_for_tests() -> None:
    """Reset store (tests only, requires GCS)."""
    users_store.clear_all_users()

File: backend/src/services/auth.service.ts
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import { randomUUID } from "crypto";
import { env } from "../config/env";
import type { HealthProfile } from "../models/HealthProfile.model";
import {
  toPublicUser,
  type PublicUser,
  type UserRecord,
} from "../models/User.model";

const usersById = new Map<string, UserRecord>();
const usersByEmail = new Map<string, UserRecord>();

const DEFAULT_HEALTH_PROFILE: HealthProfile = {
  ageRange: "25-34",
  conditions: [],
  allergies: [],
  medications: "",
};

export class AuthError extends Error {
  constructor(
    message: string,
    readonly statusCode: number
  ) {
    super(message);
    this.name = "AuthError";
  }
}

function signToken(userId: string): string {
  return jwt.sign({ sub: userId }, env.AUTH_SECRET, { expiresIn: "7d" });
}

export function verifyToken(token: string): string {
  try {
    const payload = jwt.verify(token, env.AUTH_SECRET) as { sub: string };
    return payload.sub;
  } catch {
    throw new AuthError("Invalid or expired session", 401);
  }
}

export async function signup(
  email: string,
  password: string,
  name: string
): Promise<{ user: PublicUser; token: string }> {
  const normalized = email.trim().toLowerCase();
  if (usersByEmail.has(normalized)) {
    throw new AuthError("An account with this email already exists", 409);
  }

  const passwordHash = await bcrypt.hash(password, 10);
  const user: UserRecord = {
    id: randomUUID(),
    email: normalized,
    passwordHash,
    name: name.trim(),
    healthProfile: { ...DEFAULT_HEALTH_PROFILE },
    createdAt: new Date().toISOString(),
  };

  usersById.set(user.id, user);
  usersByEmail.set(normalized, user);

  return { user: toPublicUser(user), token: signToken(user.id) };
}

export async function login(
  email: string,
  password: string
): Promise<{ user: PublicUser; token: string }> {
  const normalized = email.trim().toLowerCase();
  const user = usersByEmail.get(normalized);
  if (!user) {
    throw new AuthError("Invalid email or password", 401);
  }

  const valid = await bcrypt.compare(password, user.passwordHash);
  if (!valid) {
    throw new AuthError("Invalid email or password", 401);
  }

  return { user: toPublicUser(user), token: signToken(user.id) };
}

export function getUserById(id: string): PublicUser | null {
  const user = usersById.get(id);
  return user ? toPublicUser(user) : null;
}

export function updateUserProfile(
  id: string,
  updates: { name?: string; healthProfile?: HealthProfile }
): PublicUser {
  const user = usersById.get(id);
  if (!user) {
    throw new AuthError("User not found", 404);
  }

  if (updates.name !== undefined) {
    user.name = updates.name.trim();
  }
  if (updates.healthProfile !== undefined) {
    user.healthProfile = updates.healthProfile;
  }

  return toPublicUser(user);
}
